File: lib/sidereon/geodetic_time_series.ex
defmodule Sidereon.GeodeticTimeSeries do
  @moduledoc """
  Geodetic position time-series velocity, trajectory, step, and network tools.

  Epochs are decimal years. Positions are metres in the selected frame.
  """

  alias Sidereon.NIF

  defmodule PositionSample do
    @moduledoc """
    One position sample in a geodetic time series.
    """

    @enforce_keys [:epoch_year, :position_m]
    defstruct [:epoch_year, :position_m, :covariance_m2]

    @type t :: %__MODULE__{
            epoch_year: float(),
            position_m: {float(), float(), float()},
            covariance_m2: [[float()]] | nil
          }
  end

  defmodule MidasOptions do
    @moduledoc """
    Options for the MIDAS velocity estimator.
    """

    defstruct dominant_period_years: 1.0,
              period_tolerance_years: 0.001,
              min_pairs: 3

    @type t :: %__MODULE__{
            dominant_period_years: float(),
            period_tolerance_years: float(),
            min_pairs: pos_integer()
          }
  end

  defmodule MidasComponentStats do
    @moduledoc """
    MIDAS pair-selection diagnostics for one ENU component.
    """

    @enforce_keys [:pair_count, :retained_pair_count, :slope_sigma_m_per_yr, :effective_pair_count]
    defstruct [:pair_count, :retained_pair_count, :slope_sigma_m_per_yr, :effective_pair_count]

    @type t :: %__MODULE__{
            pair_count: non_neg_integer(),
            retained_pair_count: non_neg_integer(),
            slope_sigma_m_per_yr: float(),
            effective_pair_count: float()
          }
  end

  defmodule Velocity do
    @moduledoc """
    Robust ENU velocity estimate.
    """

    @enforce_keys [
      :rate_enu_m_per_yr,
      :sigma_enu_m_per_yr,
      :covariance_enu_m2_per_yr2,
      :component_stats,
      :sample_count,
      :span_years,
      :quality
    ]
    defstruct [
      :rate_enu_m_per_yr,
      :sigma_enu_m_per_yr,
      :covariance_enu_m2_per_yr2,
      :component_stats,
      :sample_count,
      :span_years,
      :quality
    ]

    @type t :: %__MODULE__{
            rate_enu_m_per_yr: {float(), float(), float()},
            sigma_enu_m_per_yr: {float(), float(), float()},
            covariance_enu_m2_per_yr2: [[float()]],
            component_stats: [MidasComponentStats.t()],
            sample_count: non_neg_integer(),
            span_years: float(),
            quality: :nominal | :short_span
          }
  end

  defmodule TrajectoryModel do
    @moduledoc """
    Linear trajectory model shape.
    """

    defstruct reference_epoch_year: nil,
              include_annual: true,
              include_semiannual: true,
              offset_epochs_year: []

    @type t :: %__MODULE__{
            reference_epoch_year: float() | nil,
            include_annual: boolean(),
            include_semiannual: boolean(),
            offset_epochs_year: [float()]
          }
  end

  defmodule TrajectoryFitOptions do
    @moduledoc """
    Least-squares controls for trajectory fitting.
    """

    defstruct loss: :linear, f_scale_m: 1.0, max_nfev: nil

    @type t :: %__MODULE__{
            loss: :linear | :soft_l1 | :huber | :cauchy | :arctan,
            f_scale_m: float(),
            max_nfev: pos_integer() | nil
          }
  end

  defmodule Trajectory do
    @moduledoc """
    Fitted trajectory coefficients and solver diagnostics.
    """

    defstruct [
      :reference_epoch_year,
      :terms,
      :components,
      :parameter_covariance,
      :residual_rms_enu_m,
      :geometry_quality,
      :status,
      :nfev,
      :njev,
      :cost,
      :optimality
    ]

    @type t :: %__MODULE__{reference_epoch_year: float(), terms: list(), components: list()}
  end

  defmodule StepDetectionOptions do
    @moduledoc """
    Options for displacement step detection.
    """

    defstruct window_years: 0.75,
              score_threshold: 8.0,
              min_offset_m: 1.0e-4,
              min_samples_each_side: 4,
              min_separation_years: 0.25,
              midas: %MidasOptions{}

    @type t :: %__MODULE__{
            window_years: float(),
            score_threshold: float(),
            min_offset_m: float(),
            min_samples_each_side: pos_integer(),
            min_separation_years: float(),
            midas: MidasOptions.t()
          }
  end

  defmodule StepCandidate do
    @moduledoc """
    Candidate displacement step from the labelled detector heuristic.
    """

    @enforce_keys [:epoch_year, :offset_enu_m, :score, :before_count, :after_count, :heuristic]
    defstruct [:epoch_year, :offset_enu_m, :score, :before_count, :after_count, :heuristic]

    @type t :: %__MODULE__{
            epoch_year: float(),
            offset_enu_m: {float(), float(), float()},
            score: float(),
            before_count: non_neg_integer(),
            after_count: non_neg_integer(),
            heuristic: atom()
          }
  end

  defmodule NetworkStation do
    @moduledoc """
    One station input for network motion-field estimation.
    """

    @enforce_keys [:id, :reference, :samples]
    defstruct [:id, :reference, :samples, frame: :enu]

    @type t :: %__MODULE__{
            id: String.t(),
            reference: {float(), float(), float()},
            samples: [PositionSample.t()],
            frame: Sidereon.GeodeticTimeSeries.frame()
          }
  end

  defmodule MotionField do
    @moduledoc """
    Network motion field in one output ENU frame.
    """

    @enforce_keys [:origin, :remove_common_mode, :stations, :common_mode_enu_m_per_yr]
    defstruct [:origin, :remove_common_mode, :stations, :common_mode_enu_m_per_yr]

    @type t :: %__MODULE__{
            origin: {float(), float(), float()},
            remove_common_mode: boolean(),
            stations: [map()],
            common_mode_enu_m_per_yr: {float(), float(), float()}
          }
  end

  @type frame :: :enu | {:ecef, {number(), number(), number()}}
  @type sample :: PositionSample.t() | {number(), {number(), number(), number()}}
  @type error_reason ::
          :invalid_input
          | :too_few_samples
          | :insufficient_pairs
          | :singular_trajectory
          | :did_not_converge
          | :solver
          | term()

  @doc """
  Estimate station velocity with MIDAS.

  Options accept the `MidasOptions` fields plus `:frame`, which defaults to
  `:enu`.
  """
  @spec velocity_midas([sample()], keyword() | MidasOptions.t()) :: {:ok, Velocity.t()} | {:error, error_reason()}
  def velocity_midas(samples, opts \\ []) do
    opts = options_map(opts)

    case NIF.geodetic_velocity_midas(
           frame_term(Map.get(opts, :frame, :enu)),
           sample_terms(samples),
           midas_options(opts)
         ) do
      {:ok, velocity} -> {:ok, velocity(velocity)}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Fit a linear geodetic trajectory model.

  Pass a `TrajectoryModel` as the second argument and fit options as the third.
  The `:frame` option selects `:enu` or `{:ecef, reference}` input positions.
  """
  @spec fit_trajectory([sample()], TrajectoryModel.t(), keyword() | TrajectoryFitOptions.t()) ::
          {:ok, Trajectory.t()} | {:error, error_reason()}
  def fit_trajectory(samples, model \\ %TrajectoryModel{}, opts \\ []) do
    opts = options_map(opts)

    case NIF.geodetic_fit_trajectory(
           frame_term(Map.get(opts, :frame, :enu)),
           sample_terms(samples),
           trajectory_model(model),
           trajectory_fit_options(opts)
         ) do
      {:ok, trajectory} -> {:ok, trajectory(trajectory)}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Detect displacement step candidates.
  """
  @spec detect_steps([sample()], keyword() | StepDetectionOptions.t()) ::
          {:ok, [StepCandidate.t()]} | {:error, error_reason()}
  def detect_steps(samples, opts \\ []) do
    opts = options_map(opts)

    case NIF.geodetic_detect_steps(
           frame_term(Map.get(opts, :frame, :enu)),
           sample_terms(samples),
           step_options(opts)
         ) do
      {:ok, candidates} -> {:ok, Enum.map(candidates, &step_candidate/1)}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Estimate a network motion field in one output frame.

  `origin` is `{lat_rad, lon_rad, height_m}` for the output local ENU frame.
  """
  @spec network_field([NetworkStation.t()], {number(), number(), number()}, keyword()) ::
          {:ok, MotionField.t()} | {:error, error_reason()}
  def network_field(stations, origin, opts \\ []) do
    frame = {geodetic(origin), Keyword.get(opts, :remove_common_mode, false)}

    case NIF.geodetic_network_field(Enum.map(stations, &network_station/1), frame) do
      {:ok, field} -> {:ok, motion_field(field)}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp options_map(%MidasOptions{} = opts), do: Map.from_struct(opts)
  defp options_map(%TrajectoryFitOptions{} = opts), do: Map.from_struct(opts)
  defp options_map(%StepDetectionOptions{} = opts), do: Map.from_struct(opts)
  defp options_map(opts) when is_list(opts), do: Map.new(opts)
  defp options_map(opts) when is_map(opts), do: opts

  defp sample_terms(samples), do: Enum.map(samples, &sample_term/1)

  defp sample_term(%PositionSample{} = sample) do
    {sample.epoch_year / 1.0, vec3(sample.position_m), matrix_or_nil(sample.covariance_m2)}
  end

  defp sample_term({epoch_year, position_m}) do
    {epoch_year / 1.0, vec3(position_m), nil}
  end

  defp frame_term(:enu), do: {"enu", nil}
  defp frame_term({:ecef, reference}), do: {"ecef", geodetic(reference)}

  defp midas_options(opts) do
    {
      Map.get(opts, :dominant_period_years, 1.0) / 1.0,
      Map.get(opts, :period_tolerance_years, 0.001) / 1.0,
      Map.get(opts, :min_pairs, 3)
    }
  end

  defp trajectory_model(%TrajectoryModel{} = model) do
    {
      model.reference_epoch_year,
      model.include_annual,
      model.include_semiannual,
      Enum.map(model.offset_epochs_year, &(&1 / 1.0))
    }
  end

  defp trajectory_fit_options(opts) do
    {
      Map.get(opts, :loss, :linear) |> to_string(),
      Map.get(opts, :f_scale_m, 1.0) / 1.0,
      Map.get(opts, :max_nfev)
    }
  end

  defp step_options(%{midas: %MidasOptions{} = midas} = opts) do
    step_options(%{opts | midas: Map.from_struct(midas)})
  end

  defp step_options(opts) do
    {
      Map.get(opts, :window_years, 0.75) / 1.0,
      Map.get(opts, :score_threshold, 8.0) / 1.0,
      Map.get(opts, :min_offset_m, 1.0e-4) / 1.0,
      Map.get(opts, :min_samples_each_side, 4),
      Map.get(opts, :min_separation_years, 0.25) / 1.0,
      midas_options(Map.get(opts, :midas, %{}))
    }
  end

  defp network_station(%NetworkStation{} = station) do
    {
      station.id,
      geodetic(station.reference),
      frame_term(station.frame),
      sample_terms(station.samples)
    }
  end

  defp velocity(value) do
    %Velocity{
      rate_enu_m_per_yr: value.rate_enu_m_per_yr,
      sigma_enu_m_per_yr: value.sigma_enu_m_per_yr,
      covariance_enu_m2_per_yr2: value.covariance_enu_m2_per_yr2,
      component_stats: Enum.map(value.component_stats, &component_stats/1),
      sample_count: value.sample_count,
      span_years: value.span_years,
      quality: String.to_atom(value.quality)
    }
  end

  defp component_stats(value) do
    %MidasComponentStats{
      pair_count: value.pair_count,
      retained_pair_count: value.retained_pair_count,
      slope_sigma_m_per_yr: value.slope_sigma_m_per_yr,
      effective_pair_count: value.effective_pair_count
    }
  end

  defp trajectory(value) do
    %Trajectory{
      reference_epoch_year: value.reference_epoch_year,
      terms: value.terms,
      components: value.components,
      parameter_covariance: value.parameter_covariance,
      residual_rms_enu_m: value.residual_rms_enu_m,
      geometry_quality: value.geometry_quality,
      status: value.status,
      nfev: value.nfev,
      njev: value.njev,
      cost: value.cost,
      optimality: value.optimality
    }
  end

  defp step_candidate(value) do
    %StepCandidate{
      epoch_year: value.epoch_year,
      offset_enu_m: value.offset_enu_m,
      score: value.score,
      before_count: value.before_count,
      after_count: value.after_count,
      heuristic: String.to_atom(value.heuristic)
    }
  end

  defp motion_field(value) do
    %MotionField{
      origin: value.origin,
      remove_common_mode: value.remove_common_mode,
      stations: value.stations,
      common_mode_enu_m_per_yr: value.common_mode_enu_m_per_yr
    }
  end

  defp geodetic({lat_rad, lon_rad, height_m}), do: {lat_rad / 1.0, lon_rad / 1.0, height_m / 1.0}
  defp vec3({x, y, z}), do: {x / 1.0, y / 1.0, z / 1.0}
  defp vec3([x, y, z]), do: {x / 1.0, y / 1.0, z / 1.0}
  defp matrix_or_nil(nil), do: nil
  defp matrix_or_nil(matrix), do: Enum.map(matrix, fn row -> Enum.map(row, &(&1 / 1.0)) end)
end

File: native/sidereon_nif/src/geodetic_time_series.rs
//! Rustler boundary for geodetic position time-series estimators.

use rustler::{Encoder, Env, Error, NifResult, Term};
use sidereon_core::frame::Wgs84Geodetic;
use sidereon_core::geodetic_time_series::{
    detect_steps, fit_trajectory, network_field, velocity_midas, GeodeticTimeSeriesError, Loss,
    MidasComponentStats, MidasOptions, MotionField, NetworkFrame, NetworkStation, PositionFrame,
    PositionSample, PositionSeries, StepCandidate, StepDetectionOptions, TimeSeriesQuality,
    Trajectory, TrajectoryComponent, TrajectoryFitOptions, TrajectoryModel, TrajectoryTerm,
    Velocity,
};
use sidereon_core::geometry_quality::{GeometryQuality, ObservabilityTier};

mod atoms {
    rustler::atoms! {
        ok,
        error,
        invalid_input,
        too_few_samples,
        insufficient_pairs,
        singular_trajectory,
        did_not_converge,
        solver
    }
}

type Vec3 = (f64, f64, f64);
type GeodeticTerm = (f64, f64, f64);
type SampleTerm = (f64, Vec3, Option<Vec<Vec<f64>>>);
type FrameTerm = (String, Option<GeodeticTerm>);
type MidasOptionsTerm = (f64, f64, usize);
type TrajectoryModelTerm = (Option<f64>, bool, bool, Vec<f64>);
type TrajectoryFitOptionsTerm = (String, f64, Option<usize>);
type StepOptionsTerm = (f64, f64, f64, usize, f64, MidasOptionsTerm);
type NetworkFrameTerm = (GeodeticTerm, bool);
type NetworkStationTerm = (String, GeodeticTerm, FrameTerm, Vec<SampleTerm>);

#[derive(Debug, Clone, rustler::NifMap)]
struct MidasComponentStatsTerm {
    pair_count: i64,
    retained_pair_count: i64,
    slope_sigma_m_per_yr: f64,
    effective_pair_count: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct VelocityTerm {
    rate_enu_m_per_yr: Vec3,
    sigma_enu_m_per_yr: Vec3,
    covariance_enu_m2_per_yr2: Vec<Vec<f64>>,
    component_stats: Vec<MidasComponentStatsTerm>,
    sample_count: i64,
    span_years: f64,
    quality: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct GeometryQualityTerm {
    tier: String,
    redundancy: i64,
    rank: i64,
    condition_number: f64,
    gdop: f64,
    raim_checkable: bool,
    covariance_validated: bool,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TrajectoryComponentTerm {
    position_m: f64,
    velocity_m_per_yr: f64,
    annual_sin_m: Option<f64>,
    annual_cos_m: Option<f64>,
    semiannual_sin_m: Option<f64>,
    semiannual_cos_m: Option<f64>,
    offsets_m: Vec<f64>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TrajectoryTermOut {
    reference_epoch_year: f64,
    terms: Vec<(String, Option<i64>, Option<f64>)>,
    components: Vec<TrajectoryComponentTerm>,
    parameter_covariance: Vec<Vec<f64>>,
    residual_rms_enu_m: Vec3,
    geometry_quality: GeometryQualityTerm,
    status: i32,
    nfev: i64,
    njev: i64,
    cost: f64,
    optimality: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct StepCandidateTerm {
    epoch_year: f64,
    offset_enu_m: Vec3,
    score: f64,
    before_count: i64,
    after_count: i64,
    heuristic: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct StationMotionTerm {
    id: String,
    rate_enu_m_per_yr: Vec3,
    raw_rate_enu_m_per_yr: Vec3,
    sigma_enu_m_per_yr: Vec3,
    local_velocity: VelocityTerm,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct MotionFieldTerm {
    origin: GeodeticTerm,
    remove_common_mode: bool,
    stations: Vec<StationMotionTerm>,
    common_mode_enu_m_per_yr: Vec3,
}

struct StationOwned {
    id: String,
    reference: Wgs84Geodetic,
    frame: PositionFrame,
    samples: Vec<PositionSample>,
}

fn vec3(value: [f64; 3]) -> Vec3 {
    (value[0], value[1], value[2])
}

fn matrix3(rows: Vec<Vec<f64>>) -> NifResult<[[f64; 3]; 3]> {
    if rows.len() != 3 || rows.iter().any(|row| row.len() != 3) {
        return Err(Error::Term(Box::new("matrix must be 3x3")));
    }
    Ok([
        [rows[0][0], rows[0][1], rows[0][2]],
        [rows[1][0], rows[1][1], rows[1][2]],
        [rows[2][0], rows[2][1], rows[2][2]],
    ])
}

fn matrix3_vec(matrix: [[f64; 3]; 3]) -> Vec<Vec<f64>> {
    matrix
        .into_iter()
        .map(|row| row.into_iter().collect())
        .collect()
}

fn geodetic((lat_rad, lon_rad, height_m): GeodeticTerm) -> NifResult<Wgs84Geodetic> {
    Wgs84Geodetic::new(lat_rad, lon_rad, height_m).map_err(crate::errors::invalid_input)
}

fn frame((kind, reference): FrameTerm) -> NifResult<PositionFrame> {
    match kind.as_str() {
        "enu" => Ok(PositionFrame::Enu),
        "ecef" => Ok(PositionFrame::Ecef {
            reference: geodetic(
                reference
                    .ok_or_else(|| Error::Term(Box::new("ecef frame requires a reference")))?,
            )?,
        }),
        _ => Err(Error::Term(Box::new("unknown position frame"))),
    }
}

fn sample((epoch_year, (x, y, z), covariance): SampleTerm) -> NifResult<PositionSample> {
    Ok(PositionSample {
        epoch_year,
        position_m: [x, y, z],
        covariance_m2: covariance.map(matrix3).transpose()?,
    })
}

fn samples(values: Vec<SampleTerm>) -> NifResult<Vec<PositionSample>> {
    values.into_iter().map(sample).collect()
}

fn midas_options(
    (dominant_period_years, period_tolerance_years, min_pairs): MidasOptionsTerm,
) -> MidasOptions {
    let mut options = MidasOptions::default();
    options.dominant_period_years = dominant_period_years;
    options.period_tolerance_years = period_tolerance_years;
    options.min_pairs = min_pairs;
    options
}

fn trajectory_model(
    (reference_epoch_year, include_annual, include_semiannual, offset_epochs_year): TrajectoryModelTerm,
) -> TrajectoryModel {
    TrajectoryModel {
        reference_epoch_year,
        include_annual,
        include_semiannual,
        offset_epochs_year,
    }
}

fn loss(kind: &str) -> NifResult<Loss> {
    match kind {
        "linear" => Ok(Loss::Linear),
        "soft_l1" => Ok(Loss::SoftL1),
        "huber" => Ok(Loss::Huber),
        "cauchy" => Ok(Loss::Cauchy),
        "arctan" => Ok(Loss::Arctan),
        _ => Err(Error::Term(Box::new("unknown trajectory loss"))),
    }
}

fn fit_options(
    (loss_kind, f_scale_m, max_nfev): TrajectoryFitOptionsTerm,
) -> NifResult<TrajectoryFitOptions> {
    let mut options = TrajectoryFitOptions::default();
    options.loss = loss(&loss_kind)?;
    options.f_scale_m = f_scale_m;
    options.max_nfev = max_nfev;
    Ok(options)
}

fn step_options(
    (
        window_years,
        score_threshold,
        min_offset_m,
        min_samples_each_side,
        min_separation_years,
        midas,
    ): StepOptionsTerm,
) -> StepDetectionOptions {
    let mut options = StepDetectionOptions::default();
    options.window_years = window_years;
    options.score_threshold = score_threshold;
    options.min_offset_m = min_offset_m;
    options.min_samples_each_side = min_samples_each_side;
    options.min_separation_years = min_separation_years;
    options.midas = midas_options(midas);
    options
}

fn network_frame(
    ((lat_rad, lon_rad, height_m), remove_common_mode): NetworkFrameTerm,
) -> NifResult<NetworkFrame> {
    Ok(NetworkFrame {
        origin: geodetic((lat_rad, lon_rad, height_m))?,
        remove_common_mode,
    })
}

fn quality(value: TimeSeriesQuality) -> String {
    match value {
        TimeSeriesQuality::Nominal => "nominal",
        TimeSeriesQuality::ShortSpan => "short_span",
    }
    .to_string()
}

fn component_stats(value: MidasComponentStats) -> MidasComponentStatsTerm {
    MidasComponentStatsTerm {
        pair_count: value.pair_count as i64,
        retained_pair_count: value.retained_pair_count as i64,
        slope_sigma_m_per_yr: value.slope_sigma_m_per_yr,
        effective_pair_count: value.effective_pair_count,
    }
}

fn velocity_term(value: Velocity) -> VelocityTerm {
    VelocityTerm {
        rate_enu_m_per_yr: vec3(value.rate_enu_m_per_yr),
        sigma_enu_m_per_yr: vec3(value.sigma_enu_m_per_yr),
        covariance_enu_m2_per_yr2: matrix3_vec(value.covariance_enu_m2_per_yr2),
        component_stats: value
            .component_stats
            .into_iter()
            .map(component_stats)
            .collect(),
        sample_count: value.sample_count as i64,
        span_years: value.span_years,
        quality: quality(value.quality),
    }
}

fn geometry_quality(value: GeometryQuality) -> GeometryQualityTerm {
    let tier = match value.tier {
        ObservabilityTier::RankDeficient => "rank_deficient",
        ObservabilityTier::ZeroRedundancy => "zero_redundancy",
        ObservabilityTier::Weak => "weak",
        ObservabilityTier::Nominal => "nominal",
    };
    GeometryQualityTerm {
        tier: tier.to_string(),
        redundancy: value.redundancy as i64,
        rank: value.rank as i64,
        condition_number: value.condition_number,
        gdop: value.gdop,
        raim_checkable: value.raim_checkable,
        covariance_validated: value.covariance_validated,
    }
}

fn trajectory_component(value: TrajectoryComponent) -> TrajectoryComponentTerm {
    TrajectoryComponentTerm {
        position_m: value.position_m,
        velocity_m_per_yr: value.velocity_m_per_yr,
        annual_sin_m: value.annual_sin_m,
        annual_cos_m: value.annual_cos_m,
        semiannual_sin_m: value.semiannual_sin_m,
        semiannual_cos_m: value.semiannual_cos_m,
        offsets_m: value.offsets_m,
    }
}

fn trajectory_term(value: TrajectoryTerm) -> (String, Option<i64>, Option<f64>) {
    match value {
        TrajectoryTerm::Position => ("position".to_string(), None, None),
        TrajectoryTerm::Velocity => ("velocity".to_string(), None, None),
        TrajectoryTerm::AnnualSin => ("annual_sin".to_string(), None, None),
        TrajectoryTerm::AnnualCos => ("annual_cos".to_string(), None, None),
        TrajectoryTerm::SemiannualSin => ("semiannual_sin".to_string(), None, None),
        TrajectoryTerm::SemiannualCos => ("semiannual_cos".to_string(), None, None),
        TrajectoryTerm::Offset { index, epoch_year } => {
            ("offset".to_string(), Some(index as i64), Some(epoch_year))
        }
    }
}

fn trajectory_term_out(value: Trajectory) -> TrajectoryTermOut {
    TrajectoryTermOut {
        reference_epoch_year: value.reference_epoch_year,
        terms: value.terms.into_iter().map(trajectory_term).collect(),
        components: value
            .components
            .into_iter()
            .map(trajectory_component)
            .collect(),
        parameter_covariance: value.parameter_covariance,
        residual_rms_enu_m: vec3(value.residual_rms_enu_m),
        geometry_quality: geometry_quality(value.geometry_quality),
        status: value.status,
        nfev: value.nfev as i64,
        njev: value.njev as i64,
        cost: value.cost,
        optimality: value.optimality,
    }
}

fn step_candidate(value: StepCandidate) -> StepCandidateTerm {
    StepCandidateTerm {
        epoch_year: value.epoch_year,
        offset_enu_m: vec3(value.offset_enu_m),
        score: value.score,
        before_count: value.before_count as i64,
        after_count: value.after_count as i64,
        heuristic: "detrended_sliding_median".to_string(),
    }
}

fn station_owned(
    (id, reference, series_frame, series_samples): NetworkStationTerm,
) -> NifResult<StationOwned> {
    Ok(StationOwned {
        id,
        reference: geodetic(reference)?,
        frame: frame(series_frame)?,
        samples: samples(series_samples)?,
    })
}

fn station_motion(value: sidereon_core::geodetic_time_series::StationMotion) -> StationMotionTerm {
    StationMotionTerm {
        id: value.id,
        rate_enu_m_per_yr: vec3(value.rate_enu_m_per_yr),
        raw_rate_enu_m_per_yr: vec3(value.raw_rate_enu_m_per_yr),
        sigma_enu_m_per_yr: vec3(value.sigma_enu_m_per_yr),
        local_velocity: velocity_term(value.local_velocity),
    }
}

fn motion_field(value: MotionField) -> MotionFieldTerm {
    MotionFieldTerm {
        origin: (
            value.frame.origin.lat_rad,
            value.frame.origin.lon_rad,
            value.frame.origin.height_m,
        ),
        remove_common_mode: value.frame.remove_common_mode,
        stations: value.stations.into_iter().map(station_motion).collect(),
        common_mode_enu_m_per_yr: vec3(value.common_mode_enu_m_per_yr),
    }
}

fn error_atom(error: GeodeticTimeSeriesError) -> rustler::Atom {
    match error {
        GeodeticTimeSeriesError::InvalidInput { .. } => atoms::invalid_input(),
        GeodeticTimeSeriesError::TooFewSamples { .. } => atoms::too_few_samples(),
        GeodeticTimeSeriesError::InsufficientPairs { .. } => atoms::insufficient_pairs(),
        GeodeticTimeSeriesError::SingularTrajectory => atoms::singular_trajectory(),
        GeodeticTimeSeriesError::DidNotConverge { .. } => atoms::did_not_converge(),
        GeodeticTimeSeriesError::Solver(_) => atoms::solver(),
    }
}

fn encode_result<'a, T: Encoder>(
    env: Env<'a>,
    result: Result<T, GeodeticTimeSeriesError>,
) -> Term<'a> {
    match result {
        Ok(value) => (atoms::ok(), value).encode(env),
        Err(error) => (atoms::error(), error_atom(error)).encode(env),
    }
}

/// Estimate robust station velocity with MIDAS.
#[rustler::nif(schedule = "DirtyCpu")]
fn geodetic_velocity_midas<'a>(
    env: Env<'a>,
    frame_term: FrameTerm,
    sample_terms: Vec<SampleTerm>,
    opts: MidasOptionsTerm,
) -> NifResult<Term<'a>> {
    let frame = frame(frame_term)?;
    let samples = samples(sample_terms)?;
    let series = PositionSeries {
        frame,
        samples: &samples,
    };
    Ok(encode_result(
        env,
        velocity_midas(&series, midas_options(opts)).map(velocity_term),
    ))
}

/// Fit a linear trajectory model to a geodetic time series.
#[rustler::nif(schedule = "DirtyCpu")]
fn geodetic_fit_trajectory<'a>(
    env: Env<'a>,
    frame_term: FrameTerm,
    sample_terms: Vec<SampleTerm>,
    model: TrajectoryModelTerm,
    opts: TrajectoryFitOptionsTerm,
) -> NifResult<Term<'a>> {
    let frame = frame(frame_term)?;
    let samples = samples(sample_terms)?;
    let model = trajectory_model(model);
    let series = PositionSeries {
        frame,
        samples: &samples,
    };
    Ok(encode_result(
        env,
        fit_trajectory(&series, &model, fit_options(opts)?).map(trajectory_term_out),
    ))
}

/// Detect displacement step candidates.
#[rustler::nif(schedule = "DirtyCpu")]
fn geodetic_detect_steps<'a>(
    env: Env<'a>,
    frame_term: FrameTerm,
    sample_terms: Vec<SampleTerm>,
    opts: StepOptionsTerm,
) -> NifResult<Term<'a>> {
    let frame = frame(frame_term)?;
    let samples = samples(sample_terms)?;
    let series = PositionSeries {
        frame,
        samples: &samples,
    };
    Ok(encode_result(
        env,
        detect_steps(&series, step_options(opts))
            .map(|items| items.into_iter().map(step_candidate).collect::<Vec<_>>()),
    ))
}

/// Estimate a station network motion field.
#[rustler::nif(schedule = "DirtyCpu")]
fn geodetic_network_field<'a>(
    env: Env<'a>,
    station_terms: Vec<NetworkStationTerm>,
    frame_term: NetworkFrameTerm,
) -> NifResult<Term<'a>> {
    let owned = station_terms
        .into_iter()
        .map(station_owned)
        .collect::<NifResult<Vec<_>>>()?;
    let stations = owned
        .iter()
        .map(|station| NetworkStation {
            id: station.id.as_str(),
            reference: station.reference,
            series: PositionSeries {
                frame: station.frame,
                samples: &station.samples,
            },
        })
        .collect::<Vec<_>>();
    Ok(encode_result(
        env,
        network_field(&stations, network_frame(frame_term)?).map(motion_field),
    ))
}
